src/main.cpp

```diff
@@ -0,0 +1,416 @@
+#include <iostream>
+#include <map>
+#include <string>
+#include <optional>
+#include <vector>
+#include <memory>
+#include <sstream>
+#include <iterator>
+
+
+
+// ----------------
+// State class
+// ----------------
+
+class State {
+public:
+    class Env {
+        void set_var(const std::string& var_name, const std::string& var_value) {
+            vars_[var_name] = var_value;
+        }
+
+        std::optional<std::string> get_var(const std::string& var_name) const { 
+            const auto it = vars_.find(var_name);
+            
+            if (it != vars_.end()) {
+                return it->second;
+            }
+
+            return {};
+
+        }
+
+    private:
+        std::map<std::string, std::string> vars_;
+    };
+
+};
+
+
+
+// ----------------
+// Streams classes
+// ----------------
+
+class IStreams {
+public:
+    virtual std::istream& get_in_stream() = 0;
+    virtual std::ostream& get_out_stream() = 0;
+    virtual std::ostream& get_err_stream() = 0;
+
+    virtual ~IStreams() = default;
+};
+
+class StreamsGlobal : public IStreams {
+public:
+    std::istream& get_in_stream() override {return std::cin; }
```

Comment:
Maybe
```
void put_on_in(const std::string&);
```
instead?

Reply:
Initial way allows to use operators `>>` and `<<`, which is very convenient. Also the argument can be smth else than string (e.g. int)